crypto/crmf/crmf_lib.c

```diff
@@ -0,0 +1,769 @@
+
+int OSSL_CRMF_MSG_set0_SinglePubInfo(OSSL_CRMF_SINGLEPUBINFO *spi,
+                                     int method, GENERAL_NAME *nm) {
+    int error = CRMF_R_NULL_ARGUMENT;
+    if (spi == NULL)
+        goto err;
+    if (method < OSSL_CRMF_PUB_METHOD_DONTCARE ||
+        method > OSSL_CRMF_PUB_METHOD_LDAP) {
+        error = ERR_R_PASSED_INVALID_ARGUMENT;
+        goto err;
+    }
+
+    ASN1_INTEGER_set(spi->pubMethod, method);
+    spi->pubLocation = nm;
+    return 1;
+
+ err:
+    CRMFerr(CRMF_F_OSSL_CRMF_MSG_SET0_SINGLEPUBINFO, error);
+    return 0;
+}
+
+int OSSL_CRMF_MSG_PKIPublicationInfo_push0_SinglePubInfo(
+                                 OSSL_CRMF_PKIPUBLICATIONINFO *pi,
+                                 OSSL_CRMF_SINGLEPUBINFO *spi) {
+    int error = ERR_R_MALLOC_FAILURE;
+    if (pi == NULL || spi == NULL) {
+        error = CRMF_R_NULL_ARGUMENT;
+        goto err;
+    }
+    if (pi->pubinfos == NULL)
+        if ((pi->pubinfos = sk_OSSL_CRMF_SINGLEPUBINFO_new_null()) == NULL)
+            goto err;
+
+    if (!(sk_OSSL_CRMF_SINGLEPUBINFO_push(pi->pubinfos, spi)))
+        goto err;
+    return 1;
+
+ err:
+    CRMFerr(CRMF_F_OSSL_CRMF_MSG_PKIPUBLICATIONINFO_PUSH0_SINGLEPUBINFO, error);
+    return 0;
+}
+
+int OSSL_CRMF_MSG_set_PKIPublicationInfo_action(
+                                 OSSL_CRMF_PKIPUBLICATIONINFO *pi, int action) {
+    int error = CRMF_R_NULL_ARGUMENT;
+    if (pi == NULL)
+        goto err;
+    if (action < OSSL_CRMF_PUB_ACTION_DONTPUBLISH ||
+        action > OSSL_CRMF_PUB_ACTION_PLEASEPUBLISH) {
+        error = ERR_R_PASSED_INVALID_ARGUMENT;
+        goto err;
+    }
+
+    ASN1_INTEGER_set(pi->action, action);
+    return 1;
+
+ err:
+    CRMFerr(CRMF_F_OSSL_CRMF_MSG_SET_PKIPUBLICATIONINFO_ACTION, error);
+    return 0;
+}
+
+ /* id-regCtrl-pkiPublicationInfo Control (section 6.3) */
+IMPLEMENT_CRMF_CTRL_FUNC(pkiPublicationInfo, OSSL_CRMF_PKIPUBLICATIONINFO,
+                         regCtrl)
+
+ /* id-regCtrl-oldCertID Control (section 6.5) from the given */
+IMPLEMENT_CRMF_CTRL_FUNC(oldCertID, OSSL_CRMF_CERTID, regCtrl)
+
+OSSL_CRMF_CERTID *OSSL_CRMF_CERTID_gen(const X509_NAME *issuer,
+                                       const ASN1_INTEGER *serial)
+{
+    OSSL_CRMF_CERTID *cid = NULL;
+
+    if (issuer == NULL || serial == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_CERTID_GEN, CRMF_R_NULL_ARGUMENT);
+        return NULL;
+    }
+
+    if ((cid = OSSL_CRMF_CERTID_new()) == NULL)
+        goto oom;
+
+    if (!X509_NAME_set(&cid->issuer->d.directoryName, (X509_NAME *)issuer))
+        goto oom;
+    cid->issuer->type = GEN_DIRNAME;
+
+    ASN1_INTEGER_free(cid->serialNumber);
+    if ((cid->serialNumber = ASN1_INTEGER_dup(serial)) == NULL)
+        goto oom;
+
+    return cid;
+
+ oom:
+    CRMFerr(CRMF_F_OSSL_CRMF_CERTID_GEN, ERR_R_MALLOC_FAILURE);
+    OSSL_CRMF_CERTID_free(cid);
+    return NULL;
+}
+
+ /*
+  * id-regCtrl-protocolEncrKey Control (section 6.6) */
```

Comment:
Someone has hesitate between single or multiple lines comments ? :)

Reply:
I guess this was due to a glitch in automatic re-formatting ;)